src/dmd/objc_glue.d

```diff
@@ -38,273 +37,353 @@ import dmd.backend.type;
+        }
+    }
 
-    // also wipe out segment numbers
-    for (int s = 0; s < ObjcSegment.SEG_MAX; ++s)
-        objc_segList[s] = 0;
+    override void generateModuleInfo()
+    {
+        if (Symbols.hasSymbols)
+            Symbols.getModuleInfo();
+    }
 }
 
-Symbol *objc_getCString(const(char)* str, size_t len, const(char)* symbolName, ObjcSegment segment)
+struct Segments
 {
-    objc_hasSymbols = true;
-
-    // create data
-    scope dtb = new DtBuilder();
-    dtb.nbytes(cast(uint)(len + 1), str);
-
-    // find segment
-    int seg = objc_getSegment(segment);
-
-    // create symbol
-    Symbol *s;
-    s = symbol_name(symbolName, SCstatic, type_allocn(TYarray, tstypes[TYchar]));
-    s.Sdt = dtb.finish();
-    s.Sseg = seg;
-    return s;
-}
+    enum Id
+    {
+        cString,
+        imageInfo,
+        methodName,
+        moduleInfo,
+        selectorRefs
+    }
 
-Symbol *objc_getMethVarName(const(char)* s, size_t len)
-{
-    objc_hasSymbols = true;
+    private
+    {
+        __gshared static int[segmentData.length] segments;
+
+        __gshared static Segments[__traits(allMembers, Id).length] segmentData = [
+            Segments("__cstring", "__TEXT", S_CSTRING_LITERALS),
+            Segments("__objc_imageinfo", "__DATA", S_REGULAR | S_ATTR_NO_DEAD_STRIP),
+            Segments("__objc_methname", "__TEXT", S_CSTRING_LITERALS),
+            Segments("__objc_classlist", "__DATA", S_REGULAR | S_ATTR_NO_DEAD_STRIP),
+            Segments("__objc_selrefs", "__DATA", S_ATTR_NO_DEAD_STRIP | S_LITERAL_POINTERS)
+        ];
+
+        const(char)* sectionName;
+        const(char)* segmentName;
+        int flags;
+        int alignment = 3;
+    }
 
-    StringValue *sv = objc_smethVarNameTable.update(s, len);
-    Symbol *sy = cast(Symbol *) sv.ptrvalue;
-    if (!sy)
+    static int opIndex(Id id)
     {
-        __gshared size_t classnamecount = 0;
-        char[42] namestr;
-        sprintf(namestr.ptr, "L_OBJC_METH_VAR_NAME_%lu", classnamecount++);
-        sy = objc_getCString(s, len, namestr.ptr, ObjcSegment.SEGmethname);
-        sv.ptrvalue = sy;
+        auto segmentsPtr = segments.ptr;
+
+        if (segmentsPtr[id] != 0)
+            return segmentsPtr[id];
+
+        foreach (i, seg ; segmentData)
+        {
+            version (OSX)
+            {
+                segmentsPtr[i] = MachObj.getsegment(
+                    seg.sectionName,
+                    seg.segmentName,
+                    seg.alignment,
+                    seg.flags
+                );
+            }
+            else
+            {
+                // This should never happen. If the platform is not OSX an error
+                // should have occurred sooner which should have prevented the
+                // code from getting here.
+                assert(0);
+            }
+        }
+
+        return segmentsPtr[id];
     }
-    return sy;
 }
 
-Symbol *objc_getMethVarName(Identifier *ident)
+struct Symbols
```

Comment:
Since `Symbol` is already a type meaning in dmd (`Symbols` being an `Array!Symbol` by convention), maybe call this `ObjcSymbols`?

Reply:
Probably ok though as it's private, just a note to say that it could have a grep friendly name.